### PR TITLE
In renderer if ssr rendering fails, fall back to spa rendering with warning

### PR DESCRIPTION
**Pull request by juno-yu**
Background:
creating a module as in https://codesandbox.io/s/qkxvnmv104 to take params for general redirect.

Under universal mode generate mode,
`let APP = await this.bundleRenderer.renderToString(context)` would fail, 
as bundleRenderer comes from `{ createBundleRenderer } from 'vue-server-renderer'`, in which having assumption that the route is having a template finally .

more background:
https://github.com/nuxt-community/redirect-module/issues/3

### CODE REVIEW

File: lib/core/renderer.js [hunk 1]
```diff
@@ -341,8 +341,18 @@ export default class Renderer {
...
+    if (this.noSSR || spa) {
+      return spaRenderer()
+    }
+
     // Call renderToString from the bundleRenderer and generate the HTML (will update the context as well)
-    let APP = await this.bundleRenderer.renderToString(context)
+    let APP
+    try {
+      APP = await this.bundleRenderer.renderToString(context)
+    } catch (_) {
```
Comment (pi0) [inline lib/core/renderer.js]: The real error shouldn't be thrown away. It makes debugging much harder.

File: lib/core/renderer.js [hunk 2]
```diff
@@ -306,7 +306,7 @@ export default class Renderer {
     const spa = context.spa || (context.res && context.res.spa)
     const ENV = this.options.env
 
-    if (this.noSSR || spa) {
+    const spaRenderer = async () => {
```
Comment (pi0) [inline lib/core/renderer.js]: Good, refactor idea. But why it should be async? Also, can we rename it to `renderSpa`?

File: lib/core/renderer.js [hunk 3]
```diff
@@ -341,8 +341,18 @@ export default class Renderer {
...
+      return spaRenderer()
+    }
+
     // Call renderToString from the bundleRenderer and generate the HTML (will update the context as well)
-    let APP = await this.bundleRenderer.renderToString(context)
+    let APP
+    try {
+      APP = await this.bundleRenderer.renderToString(context)
+    } catch (_) {
+      consola.warn(`ssrRenderer fails, fall back to spaRenderer for path ${url}.`)
```
Comment (pi0) [inline lib/core/renderer.js]: I consider this a breaking change. Because Nuxt behaviour regarding SSR errors will be completely different

**Review by pi0 — COMMENTED** [review]
Good path. But we need to assume more aspects. I'm going to work on a complete renderer refactor. Will take your PR into the account.

**Comment by juno-yu** [comment]
Sure, and a more thorough solution for my proposed case might be to get the route considered by vue-server-renderer. Maybe let's get this closed first

**Pull request closed without merging**